src/functional/ffront/foast_passes/type_deduction.py

```diff
@@ -0,0 +1,143 @@
+            warnings.warn(
+                FieldOperatorTypeDeductionError.from_foast_node(
+                    node,
+                    msg=(
+                        "type inconsistency: expression was deduced to be "
+                        f"of type {new}, instead of the expected type {old}"
+                    ),
+                )
+            )
+    elif isinstance(old, foast.DeferredSymbolType) and old.constraint is not None:
+        if not isinstance(new, old.constraint) or isinstance(new, foast.DeferredSymbolType):
+            warnings.warn(
+                FieldOperatorTypeDeductionError.from_foast_node(
+                    node,
+                    msg=(
+                        "type inconsistency: expression was deduced to be "
+                        f"of type {new}, instead a {type(old)} type was expected."
+                    ),
+                )
+            )
+
+
+class FieldOperatorTypeDeduction(NodeTranslator):
+    """Deduce and check types of FOAST expressions and symbols."""
+
+    contexts = (SymbolTableTrait.symtable_merger,)
+
+    @classmethod
+    def apply(cls, node: foast.FieldOperator) -> foast.FieldOperator:
+        return cls().visit(node)
+
+    def visit_FieldOperator(self, node: foast.FieldOperator, **kwargs) -> foast.FieldOperator:
+        return foast.FieldOperator(
+            id=node.id,
+            params=self.visit(node.params, **kwargs),
+            body=self.visit(node.body, **kwargs),
+            location=node.location,
+        )
+
+    def visit_Name(self, node: foast.Name, **kwargs) -> foast.Name:
+        symtable = kwargs["symtable"]
+        if node.id not in symtable:
+            warnings.warn(  # TODO (ricoh): raise this instead (requires externals)
+                FieldOperatorTypeDeductionError.from_foast_node(
+                    node, msg=f"Undeclared symbol {node.id}"
+                )
+            )
+            return node
+
+        symbol = symtable[node.id]
+        return foast.Name(id=node.id, type=symbol.type, location=node.location)
+
+    def visit_Assign(self, node: foast.Assign, **kwargs) -> foast.Assign:
+        new_value = node.value
+        if not is_complete_type(node.value.type):
+            new_value = self.visit(node.value, **kwargs)
+        new_target = self.visit(node.target, refine_type=new_value.type, **kwargs)
+        return foast.Assign(target=new_target, value=new_value, location=node.location)
+
+    def visit_FieldSymbol(
+        self, node: foast.FieldSymbol, refine_type: Optional[foast.FieldType] = None, **kwargs
+    ) -> foast.FieldSymbol:
+        symtable = kwargs["symtable"]
+        if refine_type:
+            check_type_refinement(node, refine_type)
+            new_node = foast.FieldSymbol(id=node.id, type=refine_type, location=node.location)
+            symtable[new_node.id] = new_node
+            return new_node
+        return node
+
+
+class FieldOperatorTypeDeductionError(GTSyntaxError, SyntaxWarning):
```

Comment:
I would create a new `GTSyntaxWarning` class and inherit from it

```suggestion
class FieldOperatorTypeDeductionError(GTSyntaxWarning):
```

Reply:
I would not, because this inheriting from `SyntaxWarning` is only a temporary measure until external symbols can be type checked.

Reply:
Other than that I don't have any concrete plans for a syntax warning and although I suspect it will come up in the future, let's put it in only when it is used somewhere.

Reply:
Ok, I understand, agreed.